### ISSUE TITLE
Database cleanup of orphaned records

### ISSUE THREAD

**Issue by pstorch** [cgeo/cgeo#6226]
Follow up from https://github.com/cgeo/cgeo/issues/6186#issuecomment-270253650.
It turned out that due to the lack of foreign keys in the DB it happens that geocaches got deleted from `cg_caches` without their related records in dependent tables like `cg_logs`, `cg_waypoints`, ...

The `DataStore.cleanIfNeeded()` is the housekeeping method which regulary (on startup) cleans up the database. This should be extended to also clean up these orphaned records as well.

**Comment by SammysHP** [comment]
AFAIK sqlite supports `ON DELETE CASCADE`. That might be an option (if we clean up our database scheme a little bit).

**Comment by kumy** [comment]
+1 for cleaning, adding constraints on an "unclean" database will fail. As we have lots of users, we should take care of a clean way of upgrading db, I imagine there will be cases of inconsistencies that we didn't thought about...
Good luck :)

**Comment by pstorch** [comment]
So, it's about cleaning up:

- `cg_attributes`
- `cg_spoilers`
- `cg_caches_lists`
- `cg_waypoints`
- `cg_trackables`
- `cg_logs` and cascading `cg_logimages`
- `cg_logcount`
- `cg_logs_offline`

**Comment by pstorch** [comment]
:anguished: adding foreign key constraints to sqlite is not straight forward. Need to recreate the tables one by one: https://www.sqlite.org/faq.html#q11

**Comment by pstorch** [comment]
Maybe I first add the cleanup code and later the schema migration code to add FKs. The cleanup code is needed anyway.

**Comment by Lineflyer** [comment]
AFAIR there is some code as well which checks on version upgrade whether the DB has inconsistencies and resets it completely in case of unresolvable conflicts. 
From my non-programming perspective I would therefore just like to know:
It there any probability that this change will in any way conflict with that other code (if existing?!) and is this other code (if existing?!) still needed if we have a good clean up here?

After each version upgrade we normally have like 1-2 users contacting us, saying that after upgrade all saved caches were gone. I assumed in these cases, that the above mentioned hard reset was performed as the DB had inconsistencies.

**Comment by pstorch** [comment]
There are two places where cleanup happens:
- `onUpgrade()` called on version upgrade
Here I deleted the statement `"DELETE FROM " + dbTableCaches + " WHERE reason = 0"`. This deleted caches which were not on any list. But the `reason` column is the old way for the single list DB version. I think this does the harm for the 1-2 users you are mentioning. The other statements there do explizit version upgrades from one DB version to a newer DB version. 
- `cleanIfNeeded()` called on startup and running in the background
It does the 3 day cleanup (incl. filesystem). There I added the code to cleanup possible inconsistencies.

So with this change the inconsistencies are only cleaned up in one place.
